crates/ruff/src/checkers/ast/traits.rs

```diff
@@ -0,0 +1,42 @@
+use ruff_diagnostics::Diagnostic;
+
+use crate::checkers::ast::RuleContext;
+use crate::registry::Rule;
+use crate::settings::Settings;
+
+/// Trait for a lint rule that can be run on an AST node of type `T`.
+pub(crate) trait Analyzer<T>: Sized {
+    /// The [`Rule`] that this analyzer implements.
+    fn rule() -> Rule;
+
+    /// Run the analyzer on the given node.
+    fn run(diagnostics: &mut Vec<Diagnostic>, checker: &RuleContext, node: &T);
+}
+
+/// Internal representation of a single [`Rule`] that can be run on an AST node of type `T`.
+pub(super) struct RegisteredRule<T> {
+    rule: Rule,
+    run: Executor<T>,
+}
+
+impl<T> RegisteredRule<T> {
+    pub(super) fn new<R: Analyzer<T> + 'static>() -> Self {
+        Self {
+            rule: R::rule(),
+            run: R::run,
+        }
+    }
+
+    #[inline]
+    pub(super) fn enabled(&self, settings: &Settings) -> bool {
+        settings.rules.enabled(self.rule)
+    }
+
+    #[inline]
+    pub(super) fn run(&self, diagnostics: &mut Vec<Diagnostic>, context: &RuleContext, node: &T) {
+        (self.run)(diagnostics, context, node);
+    }
```

Comment:
One unsolved problem is that we have some "lint rule functions" that add diagnostics for multiple rules in a single pass, like `rules/flake8_annotations/rules.rs`.

Reply:
`abstract_base_class` is a concrete example: it's a single function that adds violations for both `AbstractBaseClassWithoutAbstractMethod` and `EmptyMethodWithoutAbstractDecorator`, which have intertwined logic (they both rely on assessing whether a class has an abstract method).

We could break those out into separate rules and just accept that we're doing re-computation between them. Or we could make this trait one-to-many, such that a single Analyzer can produce violations for multiple rules.